### PR TITLE
Define Widget API

### PR DESCRIPTION
**Pull request by elenatorro**
Related with #809 

- [x] Widget class
- [x] Default Widget (only displays info)
- [x] Formula Widget
- [x] Multiple widgets per layer
- [x] Example notebook
- [x] Tests

Extra:
- [ ] Static vs Interactive widgets ? => Different issue

Screenshot 2019-06-26 at 12 20 19
Screenshot 2019-06-26 at 12 20 12
Screenshot 2019-06-26 at 12 20 03

### CODE REVIEW

File: cartoframes/viz/layer.py [hunk 1]
```diff
@@ -149,3 +160,34 @@ def _set_legend(legend):
...
+        return list(map(_set_widgets, widgets))
+    if isinstance(widgets, dict):
+        return Widget(widgets)
+    elif isinstance(widgets, Widget):
+        return widgets
+    else:
+        return Widget()
+
+
+def _get_widgets_variables(widgets):
```
Comment (Jesus89) [inline cartoframes/viz/layer.py]: We can move this method to the `Widget` class

File: cartoframes/viz/layer.py [hunk 2]
```diff
@@ -149,3 +160,34 @@ def _set_legend(legend):
...
+    variables = {}
+    if isinstance(widgets, list):
+        for widget in widgets:
+            if (widget._type != 'default'):
```
Comment (Jesus89) [inline cartoframes/viz/layer.py]: Paracentesis are not required here

File: cartoframes/viz/widget.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                'name': self._name,
+                'value': self._value,
+                'title': self._title,
+                'description': self._description,
+                'footer': self._footer
+            }
+        else:
+            return {}
+
+    def get_variables(self):
```
Comment (Jesus89) [inline cartoframes/viz/widget.py]: This method is not used

File: cartoframes/viz/widget.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        return self._get_vars()
+
+    def _check_type(self):
+        if self._type and self._type not in constants.WIDGET_TYPES:
+            raise ValueError(
+                'Widget type is not valid. Valid widget types are: {}.'.format(
+                    ', '.join(constants.WIDGET_TYPES)
+                ))
+
+    def _get_vars(self):
```
Comment (Jesus89) [inline cartoframes/viz/widget.py]: This method is not used

File: cartoframes/viz/layer.py [hunk 5]
```diff
@@ -149,3 +160,34 @@ def _set_legend(legend):
         return legend
     else:
         return Legend()
+
+
+def _set_widgets(widgets):
+    """Set a Widget class from the input"""
+    if isinstance(widgets, list):
+        return list(map(_set_widgets, widgets))
```
Comment (Jesus89) [inline cartoframes/viz/layer.py]: We can create a `Widgets` class to manage the list case, so we can clean the code a bit (`this.widgets.get_variables()`, etc) and move `_get_widgets_variables` and `_get_widget_info` from the Layer class

File: cartoframes/assets/templates/viz/widgets.js.j2 [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  const element = document.querySelector(`#layer${layerIndex}_widget${widgetIndex}-${widget.name}`);
+
+  if (widget.type === 'formula') {
+    element.innerText = formatToUnits(variable.value);
+  } else {
+    element.innerText = variable ? variable.value : widget.value;
+  }
+}
+
+function formatToUnits(number, precision=1) {
```
Comment (Jesus89) [inline cartoframes/assets/templates/viz/widgets.js.j2]: We are already using a `formatValue` function for the legends and popups

File: cartoframes/assets/templates/viz/map.js.j2 [hunk 7]
```diff
@@ -88,6 +91,16 @@ function onReady() {
...
       createLegend(mapLayer, layer.legend, layers.length - index - 1);
     }
+
+    mapLayer.on('updated', () => {
+      if (layer.widgets.length) {
+        layer.widgets.forEach((widget, widgetIndex) => {
+          renderWidget(mapLayer, widget, mapViz.variables[widget.name], layers.length - index - 1, widgetIndex);
+        })
+      } else {
+        renderWidget(mapLayer, layer.widgets, mapViz.variables[layer.widgets.name], layers.length - index - 1);
```
Comment (Jesus89) [inline cartoframes/assets/templates/viz/map.js.j2]: What is this use case?

**Comment by rjimenezda** [comment]
This is awesome, my only complaint is that I think the initial designs used an airship sidebar for the  widgets instead of a panel, but we can iterate on it on future widgets

**Comment by andy-esch** [comment]
So good @elenatorro 😃 

Feels similar about sidebar vs panel. I don't know if this is in scope or not but an element that hides/shows legends would be pretty nice.

I'm also feeling a desire for similar number formatting to what's in the legends and tooltips. Here it seems like it's the readable postfixes (m for million, etc.), but other data is displayed in scientific notation. Is it easy to configure the number formatting with widgets?

**Comment by Jesus89** [comment]
@andy-esch regarding the format. This is something I noticed, and I think it's fixed now (using the same utility function) :)

File: cartoframes/viz/widget_list.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+class WidgetList(object):
+    """List
+     Args:
+        widgets (dict, list, Widget): The list of widgets for a layer.
+
+    Example:
+     .. code::
+         from cartoframes.viz import Widget
+
+         WidgetList({
```
Comment (Jesus89) [inline cartoframes/viz/widget_list.py]: Missing `[ ]` in the example

File: cartoframes/viz/widget_list.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        }, {
+            'type': 'default',
+            'value': '"Custom Info"',
+        })
+    """
+
+    def __init__(self, widgets=None):
+        self.widgets = self._initwidgets(widgets)
+
+    def _initwidgets(self, widgets):
```
Comment (Jesus89) [inline cartoframes/viz/widget_list.py]: _init_widgets

**Comment by elenatorro** [comment]
@andy-esch we're using scientific notation in all the elements now, I didn't update the screenshots. Also, what you mention about panels, I think it'd be a good idea to try it once the current milestones are finished (or at least the ones related with the widgets, helpers etc), we can discuss it in the next weekly 👌

**Pull request merged**